### PR TITLE
feat: for each loops

### PR DESCRIPTION
**Pull request by zolamk**
## Description
Introduces support for `forEach` loops in reviewpad DSL.

### Summary generated by Reviewpad on 31 May 23 10:28 UTC
This pull request contains several changes across the codebase. The `env.go` and `typeinfer.go` files have modifications to support built-ins and check for their existence. The `engine` package has several changes, including new imports, implementations of built-in functions, and modifications to structs and methods to support new features. A new file `reviewpad_with_conditional_foreach_workflow.yml` has been added, which defines a group and workflow, including a conditional statement. Another new file `reviewpad_with_foreach_workflow.yml` defines a workflow that iterates over a list of teams and adds a label to each. The `lintShadowedVariables()` function has been added to handle variable shadowing in workflow runs and workflows. Finally, the `inline_rules_normalized.go` and `exec.go` files have changes to support forEach blocks and new package imports.
 

### <samp>🤖 Generated by Copilot at eb1dd26</samp>

This pull request adds support for `forEach` clauses in run blocks, using the `lang` package from reviewpad/v4 to handle list expressions and values. It updates the engine, the linter, and the aladino interpreter to parse, validate, normalize, transform, and execute the `forEach` clauses. It also adds a `Type` method to the `Value` interface and its implementations, and enables type inference and evaluation for list expressions and temporary variables in the aladino language. It modifies several files in the `engine` and `lang/aladino` packages, and refactors some test functions in `lang/aladino/interpreter_internal_test.go`.

## Code review and merge strategy

**Ask**: this pull request requires a code review before merge

### <samp>🤖 Generated by Copilot at eb1dd26</samp>

*  Import `lang` package from reviewpad/v4 to use `lang.Value` interface and other types ([link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-bd383b1ff56f0a02ff20cb2df318417891c9954441067394d8ad3ec8a30638c2R13), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-a265bb7df9128e988c6354a6c19295529954daecf10338751f3d394e0298185fR15), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-097124149197567e1cf6684921ce58756d78c728779094d2491510b7a4cb8becR18))
* Add `ProcessList` and `StoreTemporaryVariable` methods to `Interpreter` interface to evaluate and store list expressions and temporary variables in engine environment ([link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-bd383b1ff56f0a02ff20cb2df318417891c9954441067394d8ad3ec8a30638c2R47-R48))
* Enable evaluation and type inference for registered values in aladino interpreter, by checking the register map before looking for built-in functions and types ([link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-9dc9ed0b34b134c31282298a5a9a1846b3953a2b8b6dc4b324c764803aec1f59R49-R54), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-0a3c4561df0b5cd2a377ca5558cefa5642b0cdfbdeccd0611c510fdc18a6da00R120-R126))
* Implement `Type` method for all value types in aladino interpreter, returning the corresponding type instances ([link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-bded31e347eb660e9dbac95f76d462222ce4524fe5626eefefaa2482a7b349f9R51-R54), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-bded31e347eb660e9dbac95f76d462222ce4524fe5626eefefaa2482a7b349f9R84-R87), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-bded31e347eb660e9dbac95f76d462222ce4524fe5626eefefaa2482a7b349f9R114-R117), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-bded31e347eb660e9dbac95f76d462222ce4524fe5626eefefaa2482a7b349f9R144-R147), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-bded31e347eb660e9dbac95f76d462222ce4524fe5626eefefaa2482a7b349f9R187-R195), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-bded31e347eb660e9dbac95f76d462222ce4524fe5626eefefaa2482a7b349f9R219-R224), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-bded31e347eb660e9dbac95f76d462222ce4524fe5626eefefaa2482a7b349f9R249-R252))
* Add `ForEach` field to `PadWorkflowRunBlock` struct, holding a pointer to `PadWorkflowRunForEachBlock` struct, which has fields for `value`, `in`, and `do` of `forEach` clause ([link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-75c510fa57c2a4141549710b101cacf6718f98bbcd315480f224178a824a3701L121-R129))
* Add logic to normalize, transform, and execute `forEach` clause in run block, by validating the fields, calling `normalizeRun` and `transformRunBlock` recursively on `do` block, and iterating over the list expression and executing the `do` block for each value ([link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-ea18868fc6cfa08e95023e74934900b2af69a653d53a65a962abe70616a132f6R228-R265), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-31e0d9a2beee99e1ca91227b4cf55fc6b934b0d3bfef25a56fdf5fbf28879c92R417-R422), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-a265bb7df9128e988c6354a6c19295529954daecf10338751f3d394e0298185fL411-R412), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-a265bb7df9128e988c6354a6c19295529954daecf10338751f3d394e0298185fR418-R434))
* Add validation for run block to check if it has a `forEach` clause, and return an error if it has no `then` block, no actions, no extra actions, and no `forEach` block ([link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-dda45b7998c92f7c70e9a774157f5b9abd5b0be9437877b326064648d39964daR168), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-dda45b7998c92f7c70e9a774157f5b9abd5b0be9437877b326064648d39964daL181-R183))
* Rename `evalGroup` function to `evalList` in aladino interpreter, and update error messages and test functions to use the term list instead of group ([link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-097124149197567e1cf6684921ce58756d78c728779094d2491510b7a4cb8becL39-R40), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-097124149197567e1cf6684921ce58756d78c728779094d2491510b7a4cb8becL46-R47), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-097124149197567e1cf6684921ce58756d78c728779094d2491510b7a4cb8becL58-R61), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-a8b156f457738bc0d213990057f16c3297db4c42eed08cd5b0aa56c619e6a395L95-R95), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-a8b156f457738bc0d213990057f16c3297db4c42eed08cd5b0aa56c619e6a395L108-R108), [link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-a8b156f457738bc0d213990057f16c3297db4c42eed08cd5b0aa56c619e6a395L135-R135))
* Loop over the register map of the environment and add the types of the registered values to the built-in types map, to enable type inference for the registered values ([link](https://github.com/reviewpad/reviewpad/pull/911/files?diff=unified&w=0#diff-0ef0485c737ac47081f43d655de666e169819da6b26501fa14df7b7ba75f61e9R161-R164))

### CODE REVIEW

File: engine/transform.go [hunk 1]
```diff
@@ -11,7 +11,7 @@ import (
 
 func addDefaultsToRequestedReviewers(str string) string {
 	trimmedStr := strings.ReplaceAll(str, " ", "")
-	m := regexp.MustCompile(`\$assignReviewer\(((\[.*\])|(\$(team|group)\("[^"]*"\)))(,(\d+))?(,"([^"]*)")?\)`)
+	m := regexp.MustCompile(`\$assignReviewer\(((\[.*\])|(\$(team|group)\((?:(?:"[^"]*")|(?:\$[a-zA-Z0-9]+)\)))|(?:\$[a-zA-Z0-9]+))(,(\d+))?(,"([^"]*)")?\)`)
```
Comment (marcelosousa) [inline engine/transform.go:14]: This change is not related to this PR; please move it to another PR.

**Review by marcelosousa — CHANGES_REQUESTED** [review]

**Comment by marcelosousa** [comment]
@zolamk I need some time to go over these changes. Please bare with me.

File: engine/lang.go [hunk 2]
```diff
@@ -118,6 +118,13 @@ type PadWorkflowRunBlock struct {
...
 	Else    []PadWorkflowRunBlock
 	Actions []string
+	ForEach *PadWorkflowRunForEachBlock
+}
+
+type PadWorkflowRunForEachBlock struct {
+	Value string
+	In    string
+	Do    []PadWorkflowRunBlock
 }
```
Comment (marcelosousa) [inline engine/lang.go:128]: In the `PadWorkflow` we have the `yaml` tag but we don't have it. Why?

File: engine/exec.go [hunk 3]
```diff
@@ -408,12 +409,36 @@ func execActions(interpreter Interpreter, actions []string) (ExitStatus, error)
...
 	// if the run block was just a simple string
 	// there is no rule to evaluate, so just return the actions
-	if run.If == nil {
+	if run.ForEach == nil && run.If == nil {
 		// execute the actions
 		retStatus, err := execActions(interpreter, run.Actions)
 		return retStatus, run.Actions, err
 	}
 
+	// if the run block has a for-each clause
```
Comment (marcelosousa) [inline engine/exec.go:418]: No need for this comment.

File: engine/exec.go [hunk 4]
```diff
@@ -408,12 +409,36 @@ func execActions(interpreter Interpreter, actions []string) (ExitStatus, error)
...
+	if run.ForEach != nil {
+		var executedActions []string
+
+		// First process the list expression to be iterated
```
Comment (marcelosousa) [inline engine/exec.go:422]: No need for this comment.

File: engine/exec.go [hunk 5]
```diff
@@ -408,12 +409,36 @@ func execActions(interpreter Interpreter, actions []string) (ExitStatus, error)
...
+		value, err := interpreter.ProcessList(run.ForEach.In)
+		if err != nil {
+			return ExitStatusFailure, nil, err
+		}
+
+		for _, val := range value.(*lang.ArrayValue).Vals {
+			// store the value in the temporary variable
```
Comment (marcelosousa) [inline engine/exec.go:429]: No need for this comment.

File: engine/exec.go [hunk 6]
```diff
@@ -454,11 +479,10 @@ func execStatementBlock(interpreter Interpreter, runs []PadWorkflowRunBlock, rul
 	actions := []string{}
 	for _, run := range runs {
 		retStatus, runActions, err := execStatement(interpreter, run, rules)
+		actions = append(actions, runActions...)
 		if err != nil || retStatus == ExitStatusFailure {
-			return retStatus, nil, err
+			return retStatus, actions, err
 		}
-
-		actions = runActions
```
Comment (marcelosousa) [inline engine/exec.go:461]: I don't think understand this change. If an error is returned why would we want to return the actions?

**Review by marcelosousa — CHANGES_REQUESTED** [review]
Some minor tweaks - the big issue is that we need variable scoping.

The following program will "break":
```
groups:
  - name: test1
    spec: '["team1", "team2"]'
  - name: test2
    spec: '["team3", "team4"]'


workflows:
  - name: test
    run:
      - forEach:
          value: $teamName
          in: $group("test1")
          do: 
            - forEach:
                value: $teamName
                in: $group("test2")
                do: $addLabel($teamName)
            - $addLabel($teamName)
```

File: engine/exec.go (hunk 6 above)
Comment (zolamk) [reply engine/exec.go:461]: that is in order to return the actions executed so far, it's mostly to help with unit tests.

File: engine/exec.go [hunk 7]
```diff
@@ -408,12 +409,37 @@ func execActions(interpreter Interpreter, actions []string) (ExitStatus, error)
...
+			exitStatus, forEachActions, err := execStatementBlock(scopedInterpreter, run.ForEach.Do, rules)
+			executedActions = append(executedActions, forEachActions...)
+			if err != nil {
+				return exitStatus, executedActions, err
+			}
+		}
+
+		return ExitStatusSuccess, executedActions, nil
+	}
+
```
Comment (marcelosousa) [inline engine/exec.go:442]: /reviewpad explain

File: engine/exec.go (hunk 6 above)
Comment (marcelosousa) [reply engine/exec.go:461]: I still don't get it; you're changing the return value of actions if there is an error.

File: lang/aladino/eval.go [hunk 8]
```diff
@@ -48,6 +48,12 @@ func (v *Variable) Eval(e Env) (lang.Value, error) {
 		return val, nil
 	}
 
+	// TODO: we need to have a lint step to check if a variable name is a built-in function name
```
Comment (marcelosousa) [inline lang/aladino/eval.go:51]: This needs to be addressed somehow.

File: lang/value.go [hunk 9]
```diff
@@ -192,6 +218,12 @@ func (fVal *FunctionValue) HasKindOf(ty string) bool {
 	return fVal.Kind() == ty
 }
 
+// TODO: Implement a more reliable function type
+// that returns the actual parameters and return type
+func (fVal *FunctionValue) Type() Type {
+	return BuildFunctionType([]Type{}, BuildBoolType())
+}
+
```
Comment (marcelosousa) [inline lang/value.go:226]: This needs to be addressed - perhaps just return an error for now?

**Review by marcelosousa — CHANGES_REQUESTED** [review]
Some comments. For the TODOs, it is okay at this stage to simply open issues for follow-up to not block this PR (up to you to decide).

File: engine/exec.go (hunk 6 above)
Comment (zolamk) [reply engine/exec.go]: Yeah this change shouldn't be necessary for the unit tests either, I was confused by something else

File: lang/aladino/eval.go (hunk 8 above)
Comment (zolamk) [reply lang/aladino/eval.go:51]: I'll create an issue for this and address it in another pr

File: lang/value.go (hunk 9 above)
Comment (zolamk) [reply lang/value.go:226]: I'll create an issue for this and address it in another pr

**Review by marcelosousa — APPROVED** [review]

**Comment by zolamk** [comment]
We will fix in a separate pull request
- #922
- #923

**Pull request merged**